File: threads/src/main/java/im/threads/ThreadsLib.java
package im.threads;

import android.app.PendingIntent;
import android.content.Context;
import android.net.Uri;
import android.os.Build;
import android.text.TextUtils;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.util.ObjectsCompat;

import java.io.File;
import java.util.Map;

import cafe.adriel.androidaudioconverter.AndroidAudioConverter;
import cafe.adriel.androidaudioconverter.callback.ILoadCallback;
import im.threads.internal.Config;
import im.threads.internal.chat_updates.ChatUpdateProcessor;
import im.threads.internal.controllers.ChatController;
import im.threads.internal.controllers.UnreadMessagesController;
import im.threads.internal.helpers.FileProviderHelper;
import im.threads.internal.model.FileDescription;
import im.threads.internal.model.UpcomingUserMessage;
import im.threads.internal.useractivity.LastUserActivityTimeCounter;
import im.threads.internal.useractivity.LastUserActivityTimeCounterSingletonProvider;
import im.threads.internal.utils.FileUtils;
import im.threads.internal.utils.PrefUtils;
import im.threads.internal.utils.ThreadsLogger;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.exceptions.UndeliverableException;
import io.reactivex.plugins.RxJavaPlugins;
import io.reactivex.processors.FlowableProcessor;

public final class ThreadsLib {

    private static final String TAG = ThreadsLib.class.getSimpleName();

    private static ThreadsLib instance;

    private ThreadsLib() {
    }

    public static String getLibVersion() {
        return BuildConfig.VERSION_NAME;
    }

    public static void init(ConfigBuilder configBuilder) {
        if (instance != null) {
            throw new IllegalStateException("ThreadsLib has already been initialized");
        }
        Config.instance = configBuilder.build();
        instance = new ThreadsLib();
        PrefUtils.migrateToSeparateStorageIfNeeded();
        if (Config.instance.unreadMessagesCountListener != null) {
            Config.instance.unreadMessagesCountListener.onUnreadMessagesCountChanged(UnreadMessagesController.INSTANCE.getUnreadMessages());
            UnreadMessagesController.INSTANCE.getUnreadMessagesPublishProcessor()
                    .distinctUntilChanged()
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe(count -> Config.instance.unreadMessagesCountListener.onUnreadMessagesCountChanged(count),
                            error -> ThreadsLogger.e(TAG, "init " + error.getMessage()));
        }
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.Q) {
            try {
                AndroidAudioConverter.load(Config.instance.context, new ILoadCallback() {
                    @Override
                    public void onSuccess() {
                        ThreadsLogger.i(TAG, "AndroidAudioConverter was successfully loaded");
                    }

                    @Override
                    public void onFailure(Exception error) {
                        ThreadsLogger.e(TAG, "AndroidAudioConverter failed to load", error);
                    }
                });
            } catch (UnsatisfiedLinkError e) {
                ThreadsLogger.e(TAG, "AndroidAudioConverter failed to load (UnsatisfiedLinkError)", e);
            }
        }
        ChatController.getInstance();
        LastUserActivityTimeCounterSingletonProvider.INSTANCE.getLastUserActivityTimeCounter();
        if (RxJavaPlugins.getErrorHandler() == null) {
            RxJavaPlugins.setErrorHandler(throwable -> {
                if (throwable instanceof UndeliverableException) {
                    throwable = throwable.getCause();
                    if (throwable != null) {
                        ThreadsLogger.e(TAG, "global handler: ", throwable);
                    }
                    return;
                }
                Thread.currentThread().getUncaughtExceptionHandler()
                        .uncaughtException(Thread.currentThread(), throwable);
            });
        }
    }

    public static ThreadsLib getInstance() {
        if (instance == null) {
            throw new IllegalStateException("ThreadsLib should be initialized first with ThreadsLib.init()");
        }
        return instance;
    }

    public void initUser(UserInfoBuilder userInfoBuilder) {
        if (!Config.instance.clientIdIgnoreEnabled) {
            final String currentClientId = PrefUtils.getClientID();
            if (currentClientId != null && !ObjectsCompat.equals(currentClientId, userInfoBuilder.clientId)) {
                logoutClient(currentClientId);
            }
        } else {
            // it will only affect GPB, every time they try to init user we will delete user related data
            ChatController.getInstance().cleanAll();
        }
        PrefUtils.setAppMarker(userInfoBuilder.appMarker);
        PrefUtils.setNewClientId(userInfoBuilder.clientId);
        PrefUtils.setAuthToken(userInfoBuilder.authToken);
        PrefUtils.setAuthSchema(userInfoBuilder.authSchema);
        PrefUtils.setClientIdSignature(userInfoBuilder.clientIdSignature);
        PrefUtils.setUserName(userInfoBuilder.userName);
        PrefUtils.setData(userInfoBuilder.clientData);
        PrefUtils.setClientIdEncrypted(userInfoBuilder.clientIdEncrypted);
        ChatController.getInstance().sendInit();
        ChatController.getInstance().loadHistory();
    }

    public void applyChatStyle(ChatStyle chatStyle) {
        Config.instance.applyChatStyle(chatStyle);
    }

    /**
     * Used to stop receiving messages for user with provided clientId
     */
    public void logoutClient(@NonNull final String clientId) {
        if (!TextUtils.isEmpty(clientId)) {
            Config.instance.transport.sendClientOffline(clientId);
        } else {
            ThreadsLogger.i(getClass().getSimpleName(), "clientId must not be empty");
        }
    }

    /**
     * Used to post messages to chat as if written by client
     *
     * @return true, if message was successfully added to messaging queue, otherwise false
     */
    public boolean sendMessage(@Nullable String message, @Nullable File file) {
        Uri fileUri = file != null ? FileProviderHelper.getUriForFile(Config.instance.context, file) : null;
        return sendMessage(message, fileUri);
    }

    /**
     * Used to post messages to chat as if written by client
     *
     * @return true, if message was successfully added to messaging queue, otherwise false
     */
    public boolean sendMessage(@Nullable String message, @Nullable Uri fileUri) {
        ChatController chatController = ChatController.getInstance();
        if (!PrefUtils.isClientIdEmpty()) {
            FileDescription fileDescription = null;
            if (fileUri != null) {
                fileDescription = new FileDescription(
                        Config.instance.context.getString(R.string.threads_I),
                        fileUri,
                        FileUtils.getFileSize(fileUri),
                        System.currentTimeMillis()
                );
            }
            UpcomingUserMessage msg = new UpcomingUserMessage(fileDescription, null, null, message, false);
            chatController.onUserInput(msg);
            return true;
        } else {
            ThreadsLogger.i(getClass().getSimpleName(), "You might need to initialize user first with ThreadsLib.userInfo()");
            return false;
        }
    }

    /**
     * @return time in seconds since the last user activity
     */
    public long getSecondsSinceLastActivity() {
        LastUserActivityTimeCounter timeCounter = LastUserActivityTimeCounterSingletonProvider
                .INSTANCE.getLastUserActivityTimeCounter();
        return timeCounter.getSecondsSinceLastActivity();
    }

    public boolean isUserInitialized() {
        return !PrefUtils.isClientIdEmpty();
    }

    /**
     * @return FlowableProcessor that emits responses from WebSocket connection
     */
    public FlowableProcessor<Map<String, Object>> getSocketResponseMapProcessor() {
        return ChatUpdateProcessor.getInstance().getSocketResponseMapProcessor();
    }

    public interface PendingIntentCreator {
        PendingIntent create(Context context, String appMarker);
    }

    public interface UnreadMessagesCountListener {
        void onUnreadMessagesCountChanged(int count);
    }
}

File: threads/src/main/java/im/threads/ThreadsLib.kt
package im.threads

import android.annotation.SuppressLint
import android.app.PendingIntent
import android.content.Context
import android.net.Uri
import android.os.Build
import android.text.TextUtils
import im.threads.business.audioConverter.AudioConverter
import im.threads.business.audioConverter.callback.ILoadCallback
import im.threads.business.logger.LoggerEdna
import im.threads.business.models.FileDescription
import im.threads.business.rest.queries.BackendApi
import im.threads.business.rest.queries.DatastoreApi
import im.threads.business.utils.FileUtils.getFileSize
import im.threads.internal.Config
import im.threads.internal.chat_updates.ChatUpdateProcessor
import im.threads.internal.controllers.ChatController
import im.threads.internal.controllers.UnreadMessagesController
import im.threads.internal.helpers.FileProviderHelper
import im.threads.internal.model.UpcomingUserMessage
import im.threads.internal.useractivity.LastUserActivityTimeCounterSingletonProvider.getLastUserActivityTimeCounter
import im.threads.internal.utils.PrefUtils
import im.threads.styles.permissions.PermissionDescriptionDialogStyle
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.exceptions.UndeliverableException
import io.reactivex.plugins.RxJavaPlugins
import io.reactivex.processors.FlowableProcessor
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.async
import kotlinx.coroutines.launch
import java.io.File

@Suppress("unused")
class ThreadsLib private constructor() {
    /**
     * @return time in seconds since the last user activity
     */
    val secondsSinceLastActivity: Long
        get() {
            val timeCounter = getLastUserActivityTimeCounter()
            return timeCounter.getSecondsSinceLastActivity()
        }

    val isUserInitialized: Boolean get() = !PrefUtils.isClientIdEmpty

    /**
     * @return FlowableProcessor that emits responses from WebSocket connection
     */
    val socketResponseMapProcessor: FlowableProcessor<Map<String, Any>>
        get() = ChatUpdateProcessor.getInstance().socketResponseMapProcessor

    fun initUser(userInfoBuilder: UserInfoBuilder) {
        // it will only affect GPB, every time they try to init user we will delete user related data
        ChatController.getInstance().cleanAll()
        PrefUtils.appMarker = userInfoBuilder.appMarker
        PrefUtils.setNewClientId(userInfoBuilder.clientId)
        PrefUtils.authToken = userInfoBuilder.authToken
        PrefUtils.authSchema = userInfoBuilder.authSchema
        PrefUtils.clientIdSignature = userInfoBuilder.clientIdSignature
        PrefUtils.userName = userInfoBuilder.userName
        PrefUtils.data = userInfoBuilder.clientData
        PrefUtils.setClientIdEncrypted(userInfoBuilder.clientIdEncrypted)
        ChatController.getInstance().sendInit()
        ChatController.getInstance().loadHistory()
    }

    fun applyChatStyle(chatStyle: ChatStyle?) {
        Config.instance.applyChatStyle(chatStyle)
    }

    fun applyStoragePermissionDescriptionDialogStyle(
        dialogStyle: PermissionDescriptionDialogStyle
    ) {
        Config.instance.applyStoragePermissionDescriptionDialogStyle(dialogStyle)
    }

    fun applyRecordAudioPermissionDescriptionDialogStyle(
        dialogStyle: PermissionDescriptionDialogStyle
    ) {
        Config.instance.applyRecordAudioPermissionDescriptionDialogStyle(dialogStyle)
    }

    fun applyCameraPermissionDescriptionDialogStyle(
        dialogStyle: PermissionDescriptionDialogStyle
    ) {
        Config.instance.applyCameraPermissionDescriptionDialogStyle(dialogStyle)
    }

    /**
     * Used to stop receiving messages for user with provided clientId
     */
    fun logoutClient(clientId: String) {
        if (!TextUtils.isEmpty(clientId)) {
            Config.instance.transport.sendClientOffline(clientId)
        } else {
            LoggerEdna.info("clientId must not be empty")
        }
    }

    /**
     * Used to post messages to chat as if written by client
     *
     * @return true, if message was successfully added to messaging queue, otherwise false
     */
    fun sendMessage(message: String?, file: File?): Boolean {
        val fileUri = if (file != null) FileProviderHelper.getUriForFile(
            Config.instance.context,
            file
        ) else null
        return sendMessage(message, fileUri)
    }

    /**
     * Used to post messages to chat as if written by client
     *
     * @return true, if message was successfully added to messaging queue, otherwise false
     */
    fun sendMessage(message: String?, fileUri: Uri?): Boolean {
        val chatController = ChatController.getInstance()
        return if (!PrefUtils.isClientIdEmpty) {
            var fileDescription: FileDescription? = null
            if (fileUri != null) {
                fileDescription = FileDescription(
                    Config.instance.context.getString(R.string.threads_I),
                    fileUri,
                    getFileSize(fileUri),
                    System.currentTimeMillis()
                )
            }
            val msg = UpcomingUserMessage(
                fileDescription,
                null,
                null,
                message,
                false
            )
            chatController.onUserInput(msg)
            true
        } else {
            LoggerEdna.info("You might need to initialize user first with ThreadsLib.userInfo()")
            false
        }
    }

    interface PendingIntentCreator {
        fun create(context: Context, appMarker: String?): PendingIntent?
    }

    interface UnreadMessagesCountListener {
        fun onUnreadMessagesCountChanged(count: Int)
    }

    companion object {
        private var instance: ThreadsLib? = null
        private val coroutineScope = CoroutineScope(Dispatchers.Main)

        @JvmStatic
        fun getLibVersion() = BuildConfig.VERSION_NAME

        @SuppressLint("CheckResult")
        @JvmStatic
        fun init(configBuilder: ConfigBuilder) {
            check(instance == null) { "ThreadsLib has already been initialized" }
            val startInitTime = System.currentTimeMillis()
            Config.instance = configBuilder.build()
            instance = ThreadsLib()

            BackendApi.init(Config.instance)
            DatastoreApi.init(Config.instance)

            Config.instance.loggerConfig?.let { LoggerEdna.init(it) }

            PrefUtils.migrateMainSharedPreferences()

            Config.instance.unreadMessagesCountListener?.let { unreadMessagesCountListener ->
                UnreadMessagesController.INSTANCE.unreadMessagesPublishProcessor
                    .distinctUntilChanged()
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe(
                        { count: Int ->
                            unreadMessagesCountListener.onUnreadMessagesCountChanged(count)
                        }
                    ) { error: Throwable -> LoggerEdna.error("init ${error.message}") }

                coroutineScope.launch {
                    val task = async(Dispatchers.IO) {
                        UnreadMessagesController.INSTANCE.unreadMessages
                    }
                    unreadMessagesCountListener.onUnreadMessagesCountChanged(task.await())
                }
            }

            if (Build.VERSION.SDK_INT < Build.VERSION_CODES.Q) {
                try {
                    AudioConverter.load(
                        Config.instance.context,
                        object : ILoadCallback {
                            override fun onSuccess() {
                                LoggerEdna.info("AndroidAudioConverter was successfully loaded")
                            }

                            override fun onFailure(error: Exception) {
                                LoggerEdna.error("AndroidAudioConverter failed to load", error)
                            }
                        }
                    )
                } catch (e: UnsatisfiedLinkError) {
                    LoggerEdna.error(
                        "AndroidAudioConverter failed to load (UnsatisfiedLinkError)",
                        e
                    )
                }
            }
            ChatController.getInstance()
            getLastUserActivityTimeCounter()
            if (RxJavaPlugins.getErrorHandler() == null) {
                RxJavaPlugins.setErrorHandler { throwable: Throwable? ->
                    var throwableCause: Throwable? = null
                    if (throwable is UndeliverableException) {
                        throwableCause = throwable.cause
                        if (throwableCause != null) {
                            LoggerEdna.error("global handler: ", throwableCause)
                        }
                        return@setErrorHandler
                    }
                    throwableCause?.let {
                        Thread
                            .currentThread()
                            .uncaughtExceptionHandler
                            ?.uncaughtException(Thread.currentThread(), it)
                    }
                }
            }

            LoggerEdna.info("Lib_init_time: ${System.currentTimeMillis() - startInitTime}ms")
        }

        @JvmStatic
        fun getInstance(): ThreadsLib {
            checkNotNull(instance) { "ThreadsLib should be initialized first with ThreadsLib.init()" }
            return instance ?: ThreadsLib()
        }
    }
}
